go/engine/passphrase_recover.go

```
@@ -0,0 +1,291 @@
+			continue
+		}
+		expDevices[i] = *d.ProtExport()
+		idMap[d.ID] = d
+	}
+	id, err := mctx.UIs().ProvisionUI.ChooseDevice(mctx.Ctx(), keybase1.ChooseDeviceArg{
+		Devices:           expDevices,
+		CanSelectNoDevice: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	// No device chosen, we're going into the reset flow
+	if len(id) == 0 {
+		// Go directly to reset
+		return e.suggestReset(mctx)
+	}
+
+	mctx.Debug("user selected device %s", id)
+	selected, ok := idMap[id]
+	if !ok {
+		return fmt.Errorf("selected device %s not in local device map", id)
+	}
+	mctx.Debug("device details: %+v", selected)
+
+	// Roughly the same flow as in provisioning
+	switch selected.Type {
+	case libkb.DeviceTypePaper:
+		return e.loginWithPaperKey(mctx)
+	case libkb.DeviceTypeDesktop, libkb.DeviceTypeMobile:
+		return e.explainChange(mctx, selected)
+	default:
+		return fmt.Errorf("unknown device type: %v", selected.Type)
+	}
+}
+
+func (e *PassphraseRecover) suggestReset(mctx libkb.MetaContext) (err error) {
+	enterReset, err := mctx.UIs().LoginUI.PromptResetAccount(mctx.Ctx(), keybase1.PromptResetAccountArg{
+		Kind: keybase1.ResetPromptType_ENTER_FORGOT_PW,
+	})
+	if err != nil {
+		return err
+	}
+	if !enterReset {
+		// Cancel the engine as it successfully resulted in the user entering the reset pipeline.
+		return nil
+	}
+
+	// We are certain the user will not know their password, so we can disable that prompt.
+	eng := NewAccountReset(mctx.G(), e.arg.Username)
+	eng.skipPasswordPrompt = true
+	if err := eng.Run(mctx); err != nil {
+		return err
+	}
+
+	// We're ignoring eng.ResetPending() as we've disabled reset completion
+	return nil
+}
+
+func (e *PassphraseRecover) loginWithPaperKey(mctx libkb.MetaContext) (err error) {
+	// First log in using the paper key
+	loginEng := NewLoginWithPaperKey(mctx.G(), e.arg.Username)
+	if err := RunEngine2(mctx, loginEng); err != nil {
+		return err
+	}
+
+	if err := e.changePassword(mctx); err != nil {
```

what happens if you cancel the passphrase recovery here? shouldn't the whole thing abort instead of the logout/login attempt?

---

It does abort, Logout is just a step required to clean up after LoginWithPrivateKey.

---

If we cancel the prompt no error is returned, don't we attempt to login instead of logout?